src/Type/Php/InArrayFunctionTypeSpecifyingExtension.php

```diff
@@ -65,27 +69,44 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
 			);
 		}
 
+		// "Simple" all constant cases that always evaluate to true still need a nudge (e.g. needle 'foo' in haystack {'foo', 'bar'})
+		if ($arrayType instanceof ConstantArrayType && $needleType instanceof ConstantScalarType) {
+			foreach ($arrayType->getValueTypes() as $i => $valueType) {
+				if (!$arrayType->isOptionalKey($i) && $valueType->equals($needleType)) {
+					return $specifiedTypes;
+				}
+			}
+		}
+
+		// If e.g. needle is 'a' and haystack non-empty-array<int, 'a'> we can be sure that this always evaluates to true
+		// Belows HasOffset::isSuperTypeOf cannot deal with that since it calls ArrayType::hasOffsetValueType and that returns maybe at max
+		if ($needleType instanceof ConstantScalarType && $arrayType->isIterableAtLeastOnce()->yes() && $arrayValueType->equals($needleType)) {
+			return $specifiedTypes;
+		}
+
```

Comment:
I obviously dislike those 2 blocks but couldn't find something better for now unfortunately